ludwig/backend/ray.py

```diff
@@ -904,61 +919,85 @@ def supports_multiprocessing(self):
-
-            # The resulting column is named "value", which is a dict with two keys: "idx" and "data".
-            ds = ray.data.read_datasource(BinaryIgnoreNoneTypeDatasource(), **read_datasource_fn_kwargs)
-            df = self.df_engine.from_ray_dataset(ds)
-            # Persist the dataframe to prevent re-reading binary files on each subsequent map_objects call
-            df = self.df_engine.persist(df)
-            df["idx"] = self.df_engine.map_objects(df["value"], lambda row: int(row["idx"]))
-            df["value"] = self.df_engine.map_objects(df["value"], lambda row: row["data"])
-            df = df.rename(columns={"value": column.name})
+            try:
+                import daft
+            except ImportError:
+                raise ImportError(
+                    " daft is not installed. "
+                    "In order to download binary files (like images/audio/..) please run "
+                    "pip install ludwig[distributed]"
+                )
+
+            # Set the runner for executing Daft dataframes to a Ray cluster
+            # Prevent re-initialization errors if the runner is already set
+            # This can happen if there are 2 or more audio/image columns
+            assert ray.is_initialized(), "Ray should be initialized by Ludwig already at application start"
+            daft.context.set_runner_ray(address="auto", noop_if_initialized=True)
+
+            # Convert Dask Series to Dask Dataframe
+            df = column.to_frame(name=column.name)
+            df["idx"] = column.index
+
+            is_dask_df = is_dask_series_or_df(df, self)
+
+            with self.storage.cache.use_credentials():
+                if is_dask_df:
+                    df = daft.from_dask_dataframe(df)
+                else:
+                    df = daft.from_pandas(df)
+
+                df = df.select("idx", column.name)
+
+                # Download binary files in parallel
+                fs, _ = get_fs_and_path(sample_fname)
+                df = df.with_column(
+                    column.name,
+                    df[column.name].url.download(
+                        # Use 16 worker threads to maximize image read throughput over each partition
+                        max_worker_threads=16,
+                        # On error, replace value with a Null and just log the error
+                        on_error="null",
+                        fs=fs,
+                    ),
+                )
+
+                if map_fn is not None:
+                    df = df.with_column(column.name, df[column.name].apply(map_fn, return_dtype=daft.DataType.python()))
+
+                # Executes and convert Daft Dataframe to Dask DataFrame - note that this preserves partitioning
+                if is_dask_df:
+                    df = df.to_dask_dataframe()
+                    df = self.df_engine.persist(df)
+                else:
+                    df = df.to_pandas()
```

Comment:
Is this only needed for the Modin df engine? I think Pandas shouldn't use this codepath at all, right?

Reply:
Actually it's needed to convert from Daft back to either Dask or Pandas, so even Pandas would hit this code path